flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -209,6 +211,86 @@ class BoxValue : public AbstractBox, public AbstractArrayBox {
+  /// reflected in the address type.
+  BoxAddressValue(mlir::Value addr, llvm::ArrayRef<mlir::Value> lenParameters)
+      : AbstractBox(addr), lenParams{lenParameters.begin(),
+                                     lenParameters.end()} {
+    // Currently only accepts fir.(ref/ptr/heap)<fir.box<type>> mlir::Value for
+    // the address. This may change if we accept
+    // fir.(ref/ptr/heap)<fir.heap<type>> for scalar without length parameters.
+    assert(verify() &&
+           "BoxAddressValue requires mem ref to fir.box<fir.[heap|ptr]<type>>");
+  }
+
+  /// Get the fir.box<type> part of the address type.
+  fir::BoxType getBoxTy() const {
+    auto type = getAddr().getType();
+    return fir::dyn_cast_ptrEleTy(type).cast<fir::BoxType>();
+  }
+  /// Return the part of the address type after memory and box types. That is
+  /// the element type, maybe wrapped in a fir.array type.
+  mlir::Type getBaseTy() const {
+    return fir::dyn_cast_ptrEleTy(getBoxTy().getEleTy());
+  }
+  /// Get the scalar type related to the described entity
+  mlir::Type getEleTy() const {
+    auto type = getBaseTy();
+    if (auto seqTy = type.dyn_cast<fir::SequenceType>())
+      return seqTy.getEleTy();
+    return type;
+  }
+
+  /// Is this a Fortran pointer ?
+  bool isPointer() const {
+    return getBoxTy().getEleTy().isa<fir::PointerType>();
+  }
+  /// Is this an allocatable ?
+  bool isAllocatable() const {
+    return getBoxTy().getEleTy().isa<fir::HeapType>();
+  }
+  /// Is the entity an array or an assumed rank ?
+  bool hasRank() const { return getBaseTy().isa<fir::SequenceType>(); }
+  /// Is this an assumed rank ?
+  bool hasAssumedRank() const {
+    auto seqTy = getBaseTy().dyn_cast<fir::SequenceType>();
+    return seqTy && seqTy.hasUnknownShape();
+  }
+  /// Returns the rank of the entity. Beware that zero will be returned for
+  /// both scalars and assumed rank.
+  unsigned rank() const {
+    auto seqTy = getBaseTy().dyn_cast<fir::SequenceType>();
+    if (seqTy)
+      return seqTy.getDimension();
+    return 0;
+  }
+  /// Is this a character entity ?
+  bool isCharacter() const { return getEleTy().isa<fir::CharacterType>(); };
+  /// Is this a derived type entity ?
+  bool isDerived() const { return getEleTy().isa<fir::RecordType>(); };
+  /// Does this entity has any non deferred length parameters ?
+  bool hasNonDeferredLenParams() const { return !lenParams.empty(); }
+  /// Return the non deferred length parameters.
+  llvm::ArrayRef<mlir::Value> nonDeferredLenParams() { return lenParams; }
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &,
+                                       const BoxAddressValue &);
+  LLVM_DUMP_METHOD void dump() const { operator<<(llvm::errs(), *this); }
+
+protected:
+  /// Validate the address type form in the constructor.
+  bool verify() const;
+  /// Hold the non-deferred length parameters (both for characters and derived).
+  llvm::SmallVector<mlir::Value, 2> lenParams;
```

Comment:
See my previous comment, only length can be non deferred in allocatables and have a special semantics that justifies keeping track of them. These non deferred length parameters are not allowed to change during the lifetime of the  `BoxAddressValue`.

lbound, extents, exact type and so on is kept inside the box and may change during the lifetime of the `BoxAddressValue`.

Reply:
Because these are not memory locations: non-deferred length parameters are kept here as ssa values and not variables since they cannot change. They are lowered once and for all when lowering the scope, and then we need to keep track of them. Deferred parameters are computed at each allocation/associations. It's also more than a rule of what can be modified or not. It impacts what lengths parameters may be omitted in allocation statement. For the omitted ones, we need to used the lowered values from the specification expressions, so we need to keep track of these values after specification expression lowering (see [here](https://github.com/flang-compiler/f18-llvm-project/pull/561/files#diff-96be36dcf5ea9de4eb09e8224e069245ad5a3bd4cc8cae81d429f7e6cdaa68ddR554-R562))).

- `MutableBoxValue.lenParams` -> hold the lowered specification-expression of non-deferred parameters as ssa values (always defined).
- descriptor/MutableProperties -> hold the lengths parameters of entities in memory (only defined when the entity is *allocated*).

An alternative could be to store the lowered expression in the descriptor/mutable properties right after specification lowering, but I think it would lead to not so nice case where we write in an intent(in) descriptor just because it has non-deferred length parameters.

Plus, I think we want to offer the compiler implementer a clear view of what are is the length from a non-deferred specification expression values vs what is the length in the descriptor/memory. Take the following example:
```
 subroutine foo(c, n)
    character(n) :: c(:)
    call bar(c) ! take character(:), allocatable :: c(:) argument
end subroutine
```
A possible runtime check that a compiler implementer might want to implement after bar call is that if `c` was modified in bar, then its length should still be in line with the non-deferred length parameters it has in `foo` scope (`n`).
The implementer will need to know what is the non-deferred parameter value that was lowered in from the specification expr. Having the non-deferred length parameter value list here will make his life much easier. 

So, in the context of pointer/allocatable I think the distinction deferred/non-deferred is important. Note than when we will lower a reference to the entity, the MutableBox is raised to a "normal box" (e.g a `BoxValue` for derived types), then there is only one list of length parameters because the deferred/non-deferred semantics does not matter there.

Reply:
My earlier comment was poorly phrased. I meant LEN type parameters must be treated as computed values that can vary at runtime. "Variable" was confusingly used to mean "not a compile-time constant". My bad.

I think why I keep getting confused is the negative-orientation of this data structure. *Non-deferred* LEN parameters aren't special -- they are the rule. It is _deferred_ LEN parameters that are special. Those are the values that can have control-dependence, etc. (i.e. could be "variables" in memory).

The other point of confusion is how the enhanced descriptor is meant to be populated. I'm not clear about what sorts of LEN parameters are being placed in the enhanced descriptor or if some are being omitted. I thought all of them would be, but now I am not sure.